### PR TITLE
[runtime] Fix Mono BStr marshaling, cleanup IL generation

### PR DESCRIPTION
**Pull request by CoffeeFlux**
This PR checks in recent string marshaling work and enables a subset of tests that now run successfully. It's possible that this fixes or changes the failure on other tests, but I've only checked the StringMarshalling ones.

Status of string marshaling of this PR:
* AnsiBStr - tests pass
* BStr - tests pass
* LPStr - failure with StringBuilder
* LPTStr - immediate failure
* UTF8Str - crash with invalid malloc in StringBuilder code (??)
* LPWStr - no tests
* TBStr - no tests

Followup work:
- [ ] Fix UTF8 crash (probably corruption, the crash used to be in a free call)
- [ ] Clean up StringBuilder code (lots of issues, some caught by the tests, in progress)
- [ ] Add TBStr and LPWStr tests (in progress)
- [ ] Fix LPTStr to properly handle both ANSI and UTF16 text (or maybe change the test to align with our historical behavior? currently the test uses LPWStr on all platforms)
- [ ] Add a wall of text on string marshaling behavior for classic and netcore Mono at the start of `conv_to_icall`

### CODE REVIEW

File: src/mono/mono/metadata/cominterop.c [hunk 1]
```diff
@@ -2998,12 +3012,19 @@ mono_ptr_to_bstr (const gunichar2* ptr, int slen)
...
+#else // In Mono, historically BSTR was allocated with a guaranteed size prefix of 4 bytes regardless of platform
 		/* allocate len + 1 utf16 characters plus 4 byte integer for length*/
 		guint32 * const ret = (guint32 *)g_malloc ((slen + 1) * sizeof (gunichar2) + sizeof (guint32));
 		if (ret == NULL)
 			return NULL;
 		mono_bstr const s = (mono_bstr)(ret + 1);
-		*ret = slen * sizeof (gunichar2);
+#endif
+		// Set the length
+		*((guint32 *)s - 1) = slen * sizeof (gunichar2);
```
Comment (alexischr) [inline src/mono/mono/metadata/cominterop.c:3027]: I may be wrong here but, in case of 64-bit pointers, `*((guint32 *)s - 2)` may not be getting zeroed

**Comment by alexischr** [comment]
https://docs.microsoft.com/en-us/previous-versions/windows/desktop/automat/bstr has BSTR always have a four-byte integer for the length. Is this a "different" BSTR?

**Review by lambdageek — APPROVED** [review]
LGTM, in general.

Nits:
1. It'd be great to have a wall of text somewhere summarizing the string types and how we choose to marshal them in (netcore & classic) Mono
2. Address Alexis' `(guint32*)s - 2` comment
3. Maybe `bstr_set_length` as a little helper to collect the pointer arithmetic.

I like `conv_str_inverse`

File: src/mono/mono/metadata/cominterop.c (hunk 1 above)
Comment (alexischr) [reply src/mono/mono/metadata/cominterop.c]: It looks like this problem is still there in the new set_length function

Comment (alexischr) [reply src/mono/mono/metadata/cominterop.c]: Oh I see the switch to `g_malloc0` now

**Review by alexischr — APPROVED** [review]

**Pull request merged**